### PR TITLE
Adding stepwise regression feature to `PolynomialPathFitter`

### PR DESCRIPTION
**Pull request by nickbianco**
Fixes issue #3761

[perf-win]

### Brief summary of changes

Adds the option to use stepwise regression when fitting polynomial coefficients in `PolynomialPathFitter`. This approach fits polynomials using individual terms one at a time and keeps the terms that result in the lowest fitting error. Terms (and coefficients) are added until the path length and moment arm tolerances are achieved. 

One key change was made to `MultivariatePolynomialFunction`: coefficients equal to zero are now ignored when evaluating polynomial functions to maximize speed.

### Testing I've completed

- Ran Matlab and Python scripting examples, `examplePolynomialPathFitter.m/py`. 
- Updated the `FunctionBasedPath` set used in `exampleMocoInverse` and `exampleMocoTrack` and ran both examples.

### CHANGELOG.md (choose one)

- updated.

### Performance analysis

Platform: Windows, self-hosted runner

|                           Test Name | lhs [secs] | stderr [secs] | rhs [secs] | stderr [secs] | Speedup |
| ----------------------------------- | ---------- | ------------- | ---------- | ------------- | ------- |
|                               Arm26 |       0.35 |          0.00 |       0.36 |          0.00 |    0.97 |
|                      ellipsoid_wrap |       4.12 |          0.00 |       4.15 |          0.01 |    0.99 |
| ellipsoid_wrap_function_based_paths |       3.41 |          0.00 |       3.42 |          0.00 |    1.00 |
|                            Gait2354 |       0.42 |          0.00 |       0.42 |          0.00 |    1.01 |
|                     MocoSlidingMass |       1.48 |          0.00 |       1.48 |          0.00 |    1.00 |
|                     passive_dynamic |       5.61 |          0.00 |       5.59 |          0.00 |    1.00 |
|          passive_dynamic_noanalysis |       3.45 |          0.01 |       3.41 |          0.00 |    1.01 |
|                      RajagopalModel |       8.89 |          0.01 |       8.88 |          0.01 |    1.00 |
|                      ToyDropLanding |      13.38 |          0.01 |      13.41 |          0.00 |    1.00 |
|      ToyDropLanding_fbp_stepwisereg |      13.91 |          0.07 |      12.31 |          0.01 |    1.13 |
| ToyDropLanding_function_based_paths |      12.43 |          0.01 |      12.51 |          0.03 |    0.99 |
|            ToyDropLanding_nomuscles |       0.57 |          0.00 |       0.56 |          0.00 |    1.01 |

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3779)

### CODE REVIEW

File: OpenSim/Actuators/PolynomialPathFitter.cpp [hunk 1]
```diff
@@ -1310,6 +1274,209 @@ void PolynomialPathFitter::removeMomentArmColumns(TimeSeriesTable& momentArms,
...
+
+        // First, set the columns of 'A' for the "kept" coefficients.
+        int icol = 0;
+        for (const auto& ki : keepIndexes) {
+            A.updCol(icol++) = Afull.col(ki);
+        }
+
+        // Next, loop through all of the remaining "out" coefficients and fit 
+        // the polynomial. We will keep the coefficient that results in the
+        // smallest RMS error.
```
Comment (pepbos) [inline OpenSim/Actuators/PolynomialPathFitter.cpp:1406]: Nice, lots of comments :)

File: OpenSim/Actuators/PolynomialPathFitter.cpp [hunk 2]
```diff
@@ -1310,6 +1274,209 @@ void PolynomialPathFitter::removeMomentArmColumns(TimeSeriesTable& momentArms,
...
+        SimTK::Real bestError = SimTK::Infinity;
+        int bestIndex = -1;
+        for (const auto& oi : outIndexes) {
```
Comment (pepbos) [inline OpenSim/Actuators/PolynomialPathFitter.cpp:1409]: Will this not become extremely expensive as `A` starts getting some nonzero size? Or is computational time not so important during this phase?

**Review by pepbos — APPROVED** [review]
Nice, added a few comments, but very clear overall.

As a side note: `fitCoefficientsStepwiseRegression` does look expensive, with the leastsquares fit in a loop. That process looks a lot like Principal Component Analysis which can be done in one go using a singular value decomposition, unless I am mistaken.
But this is just a side note.

File: OpenSim/Actuators/PolynomialPathFitter.h [hunk 3]
```diff
@@ -633,6 +664,22 @@ class OSIMACTUATORS_API PolynomialPathFitter : public Object {
     static void removeMomentArmColumns(TimeSeriesTable& momentArms,
             const MomentArmMap& momentArmMap);
 
+    /**
+     * Fit to the path length and moment arm samples using all possible
+     * polynomial coefficients.
+     */
+    void fitAllCoefficients(const SimTK::Matrix& coordinates, 
+            const SimTK::Vector& b, int& order,
```
[the remark this answers is not in the document]
Comment (nickbianco) [reply OpenSim/Actuators/PolynomialPathFitter.h:672]: Done.

File: OpenSim/Actuators/PolynomialPathFitter.cpp (hunk 2 above)
Comment (nickbianco) [reply OpenSim/Actuators/PolynomialPathFitter.cpp:1409]: This ends up not being as expensive as it seems. The biggest bottleneck in the fitter is (by far) calculating path length and moment arm values, especially if you use a lot of sample points. But the time can be reduced significantly by sampling the model coordinates in an efficient way.

File: OpenSim/Actuators/PolynomialPathFitter.cpp [hunk 4]
```diff
@@ -1310,6 +1274,209 @@ void PolynomialPathFitter::removeMomentArmColumns(TimeSeriesTable& momentArms,
...
+
+    // Construct the full 'A' matrix.
+    SimTK::Vector dummyCoefficients(numCoefficients, 1.0);
+    MultivariatePolynomialFunction dummyFunction(dummyCoefficients,
+            numCoordinates, order);
+    SimTK::Matrix Afull(numTimes * (numCoordinates + 1),
+            numCoefficients, 0.0);
+    for (int itime = 0; itime < numTimes; ++itime) {
+        Afull(itime, 0, 1, numCoefficients) =
+                dummyFunction.getTermValues(
```
[the remark this answers is not in the document]
Comment (nickbianco) [reply OpenSim/Actuators/PolynomialPathFitter.cpp:1364]: A polynomial "term" is just a monomial. I almost commented on this in the previous PR, but changing `getTermValues()` would change the public API of `MultivariatePolynomialFunction` which was just release in OpenSim 4.5. So I'm a bit hesistant to do that.

File: OpenSim/Actuators/PolynomialPathFitter.cpp [hunk 5]
```diff
@@ -1310,6 +1274,209 @@ void PolynomialPathFitter::removeMomentArmColumns(TimeSeriesTable& momentArms,
     }
 }
 
+void PolynomialPathFitter::fitAllCoefficients(
+        const SimTK::Matrix& coordinates, const SimTK::Vector& b, int& order,
+        SimTK::Vector& coefficients) const {
+
+    int numTimes = coordinates.nrow();
+    int numCoordinates = coordinates.ncol();
+    while (true) {
```
[the remark this answers is not in the document]
Comment (nickbianco) [reply OpenSim/Actuators/PolynomialPathFitter.cpp:1283]: There is a ceiling on max order, in the if-statement at the bottom of the while loop. The current behavior is to keep increasing order until either 1) the tolerances are met or 2) the max order is met (this is documented).

**Comment by nickbianco** [comment]
Thanks @pepbos! I made some updates based on your comments. Unless you have any other comments, I'll merge after the tests pass.

**Pull request merged**